<a href="https://colab.research.google.com/github/MarkLee7925/Heart-Classification-UCI/blob/main/heart_LogReg.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Predict Heart Disease In Patient

## Data pre-processing

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import tensorflow as tf
import seaborn as sns

In [ ]:
# Load dataframe (df)
df = pd.read_csv('heart.csv')
X, y = df.iloc[:, :-1].values, df.iloc[:, -1].values

print(X.shape, y.shape)

(303, 13) (303,)


In [ ]:
print(df.head())

   age  sex  cp  trestbps  chol  fbs  ...  exang  oldpeak  slope  ca  thal  target
0   63    1   3       145   233    1  ...      0      2.3      0   0     1       1
1   37    1   2       130   250    0  ...      0      3.5      0   0     2       1
2   41    0   1       130   204    0  ...      0      1.4      2   0     2       1
3   56    1   1       120   236    0  ...      0      0.8      2   0     2       1
4   57    0   0       120   354    0  ...      1      0.6      2   0     2       1

[5 rows x 14 columns]


In [ ]:
df.dtypes

age           int64
sex           int64
cp            int64
trestbps      int64
chol          int64
fbs           int64
restecg       int64
thalach       int64
exang         int64
oldpeak     float64
slope         int64
ca            int64
thal          int64
target        int64
dtype: object

In [ ]:
df.describe()

age         sex          cp  ...          ca        thal      target
count  303.000000  303.000000  303.000000  ...  303.000000  303.000000  303.000000
mean    54.366337    0.683168    0.966997  ...    0.729373    2.313531    0.544554
std      9.082101    0.466011    1.032052  ...    1.022606    0.612277    0.498835
min     29.000000    0.000000    0.000000  ...    0.000000    0.000000    0.000000
25%     47.500000    0.000000    0.000000  ...    0.000000    2.000000    0.000000
50%     55.000000    1.000000    1.000000  ...    0.000000    2.000000    1.000000
75%     61.000000    1.000000    2.000000  ...    1.000000    3.000000    1.000000
max     77.000000    1.000000    3.000000  ...    4.000000    3.000000    1.000000

[8 rows x 14 columns]

In [ ]:
# Split dataframe (df) into training and test sets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, 
                                                    random_state=0)

In [ ]:
# Feature scaling
from sklearn.preprocessing import StandardScaler

sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

## Define ML model

In [ ]:
from sklearn.linear_model import LogisticRegression

# Accuracy Scoring
clf = LogisticRegression(C=0.2, penalty='l2', class_weight='balanced', random_state=0)

# ROC_AUC Scoring
# clf = LogisticRegression(C=0.48, penalty='l2', class_weight='balanced', random_state=0)

# Train model
clf.fit(X_train, y_train)


LogisticRegression(C=0.2, class_weight='balanced', dual=False,
                   fit_intercept=True, intercept_scaling=1, l1_ratio=None,
                   max_iter=100, multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=0, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

## Evaluate model

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

# Predict test set results
y_pred = clf.predict(X_test)
np.set_printoptions(precision = 2)
print(np.concatenate((y_pred.reshape(len(y_pred), 1), y_test.reshape(len(y_test), 1)), 1 ))

# Create the confusion matrix
# TN = cm[0][0]  # True Negative
# FP = cm[0][1]  # False Positive
# FN = cm[1][0]  # False Negative
# TP = cm[1][1]  # True Positive
cm = confusion_matrix(y_test, y_pred)
print(cm)
print("Test Accuracy: %.3f%% " % (accuracy_score(y_test, y_pred) * 100))

# accuracy = (TP + TN)/(TN + FP + FN + TP)
# precision = TP/(TP + FP)
# recall = TP/(TP + FN)
# f1_score = (2 * precision * recall)/(precision + recall)
print(classification_report(y_test, y_pred))

[[0 0]
 [1 1]
 [1 0]
 [0 0]
 [0 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [0 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [1 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [1 0]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 1]
 [0 0]
 [0 1]
 [1 1]
 [1 1]
 [1 0]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]]
[[21  6]
 [ 4 30]]
Test Accuracy: 83.607% 
              precision    recall  f1-score   support

           0       0.84      0.78      0.81        27
           1       0.83      0.88      0.86        34

    accuracy                           0.84        61
   macro avg       0.84      0.83      0.83        61
weighted avg       0.84      0.84      0.84        61



In [ ]:
# Apply k-Fold Cross Validation on the training set
from sklearn.model_selection import cross_val_score, RepeatedStratifiedKFold

# cv = RepeatedKFold()
cv = RepeatedStratifiedKFold(n_splits = 10, n_repeats = 3, random_state = 0)
scores = cross_val_score(clf, X_train, y_train, scoring='accuracy', cv=cv, n_jobs=-1)
# scores = cross_val_score(clf, X_train, y_train, scoring='roc_auc', cv=cv, n_jobs=-1)

# Force scores to be positive
scores = abs(scores)
print("Mean Accuracy: %.3f%%" % (np.mean(scores) * 100))
print("Standard Deviation: %.3f%%" % np.std(scores))
print("Min Accuracy: %.3f%%" % (np.min(scores) * 100))
print("Max Accuracy: %.3f%%" % (np.max(scores) * 100))

Mean Accuracy: 82.489%
Standard Deviation: 0.052%
Min Accuracy: 70.833%
Max Accuracy: 91.667%


In [ ]:
# Apply Grid Search for parameter tuning
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV

# param_dict = {'C': [0.2, 0.25, 0.3, 0.5, 0.75, 1],
#               'penalty': ['l1', 'l2'],
#               'class_weight': ['dict', 'balanced'],
#               'random_state': [0, 1]}

# Accuracy scoring
param_dict = {'C': [0.19, 0.2, 0.21],
              'penalty': ['l1', 'l2'],
              'class_weight': ['dict', 'balanced'],
              'random_state': [0, 1]}

# ROC_AUC scoring
# param_dict = {'C': [0.46, 0.47, 0.48, 0.49],
#               'penalty': ['l1', 'l2'],
#               'class_weight': ['dict', 'balanced'],
#               'random_state': [0, 1]}

grid = GridSearchCV(clf, param_dict, scoring='accuracy', cv=cv, n_jobs=-1)
# grid = GridSearchCV(clf, param_dict, scoring='roc_auc', cv=cv, n_jobs=-1)
results = grid.fit(X_train, y_train)

# Summarize results
best_accuracy = grid.best_score_
best_parameters = grid.best_params_
print("Best Accuracy: %.3f%%" % (best_accuracy * 100))
print("Best Parameters: ", best_parameters)


Best Accuracy: 82.489%
Best Parameters:  {'C': 0.2, 'class_weight': 'balanced', 'penalty': 'l2', 'random_state': 0}


#### -- Grid Search Results --

###### **With accuracy scoring:**

- Mean Accuracy: 81.117%\
Standard Deviation: 0.049%\
Test Accuracy: 85.246%\
Best Accuracy: 82.489%\
Best Parameters:  {'C': 0.2, 'class_weight': 'balanced', 'penalty': 'l2', 'random_state': 0}
- Mean Accuracy: 82.489%\
Standard Deviation: 0.052%\
Test Accuracy: 83.607%\
Best Accuracy: 82.489%\
Best Parameters:  {'C': 0.2, 'class_weight': 'balanced', 'penalty': 'l2', 'random_state': 0}
- Mean Accuracy: 82.489%\
Standard Deviation: 0.052%\
Test Accuracy: 83.607%\
Best Accuracy: 82.489%\
Best Parameters:  {'C': 0.2, 'class_weight': 'balanced', 'penalty': 'l2', 'random_state': 0}


###### **With ROC_AUC scoring:**

- Mean Accuracy: 89.888%\
Standard Deviation: 0.046%\
Test Accuracy: 83.607%\
Best Accuracy: 89.937%\
Best Parameters:  {'C': 0.5, 'class_weight': 'balanced', 'penalty': 'l2', 'random_state': 0}
- Mean Accuracy: 89.937%\
Standard Deviation: 0.045%\
Test Accuracy: 83.607%\
Best Accuracy: 89.937%\
Best Parameters:  {'C': 0.48, 'class_weight': 'balanced', 'penalty': 'l2', 'random_state': 0}
- Mean Accuracy: 89.937%\
Standard Deviation: 0.045%\
Test Accuracy: 83.607%\
Best Accuracy: 89.937%\
Best Parameters:  {'C': 0.48, 'class_weight': 'balanced', 'penalty': 'l2', 'random_state': 0}
